controllers/managedfusionoffering_controller.go

```diff
@@ -71,21 +93,170 @@ func (r *ManagedFusionOfferingReconciler) Reconcile(ctx context.Context, req ctr
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
 }
 
 // This function is a placeholder for offering plugin integration
 func pluginReconcile(r *ManagedFusionOfferingReconciler) (ctrl.Result, error) {
 	return ctrl.Result{}, nil
 }
 
+func (r *ManagedFusionOfferingReconciler) reconcileOperatorGroup() error {
+	r.Log.Info("Reconciling DF Offering Operator Group")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.ocsOperatorGroup, func() error {
+		if err := r.own(r.ocsOperatorGroup); err != nil {
+			return err
+		}
+		r.ocsOperatorGroup.Spec.TargetNamespaces = []string{r.namespace}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to create/update operatorGroup: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionOfferingReconciler) reconcileCatalogSource() error {
+	r.Log.Info("Reconciling DF Offering Source")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.ocsCatalogSource, func() error {
+		if err := r.own(r.ocsCatalogSource); err != nil {
+			return err
+		}
+		desired := templates.OCSCatalogSource.DeepCopy()
+		// A hook will be required to get catalog for a specific offering
+		desired.Spec.Image = "registry.redhat.io/redhat/redhat-operator-index:v4.12"
+		r.ocsCatalogSource.Spec = desired.Spec
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to create/update catalogSource: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionOfferingReconciler) reconcileSubscription() error {
+	r.Log.Info("Reconciling DF Offering Subscription")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.ocsSubscription, func() error {
+		if err := r.own(r.ocsSubscription); err != nil {
+			return err
+		}
+		desired := templates.OCSSubscription.DeepCopy()
+		desired.Spec.CatalogSource = r.ocsCatalogSource.Name
+		desired.Spec.CatalogSourceNamespace = r.ocsCatalogSource.Namespace
+		// A hook will be required to get channel and startingCSV for a specific offering
+		desired.Spec.Channel = "stable-4.12"
+		desired.Spec.StartingCSV = "ocs-operator.v4.12.0"
+		r.ocsSubscription.Spec = desired.Spec
+		return nil
+
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to create/update subscription: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionOfferingReconciler) reconcileInstallPlan() error {
+	r.Log.Info("Reconciling DF Operator InstallPlan")
+	ocsOperatorCSV := &opv1a1.ClusterServiceVersion{}
+	ocsOperatorCSV.Name = r.ocsSubscription.Spec.StartingCSV
+	ocsOperatorCSV.Namespace = r.namespace
+	if err := r.get(ocsOperatorCSV); err != nil {
+		if errors.IsNotFound(err) {
+			var foundInstallPlan bool
+			installPlans := &opv1a1.InstallPlanList{}
+			if err := r.list(installPlans); err != nil {
+				return err
+			}
+			for i, installPlan := range installPlans.Items {
+				if findInSlice(installPlan.Spec.ClusterServiceVersionNames, ocsOperatorCSV.Name) {
+					foundInstallPlan = true
+					if installPlan.Spec.Approval == opv1a1.ApprovalManual &&
+						!installPlan.Spec.Approved {
```

Comment:
This is not good enough. It will approve updates to ANY z release available in the catalog and will not lock us to the desired z release.

Reply:
No, it won't
We approve the install plan that matches the CSV name. CSV name has the z-stream mentioned.
https://github.com/red-hat-storage/managed-fusion-agent/pull/12/files#diff-15ecb6b6ffa838b8c6cbc94f2a8d7ecd88e65ed59e0f0e818cca80b0af43be07R224